File: types/sequence_chunker_test.go
package types

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

type modBoundaryChecker struct {
	mod int
}

func (b modBoundaryChecker) Write(item sequenceItem) bool {
	return item.(int)%b.mod == 0
}

func (b modBoundaryChecker) WindowSize() int {
	return 1
}

func listFromInts(ints []int) List {
	vals := make([]Value, len(ints))
	for i, v := range ints {
		vals[i] = Number(v)
	}

	return NewList(vals...)
}

func TestSequenceChunkerMod(t *testing.T) {
	assert := assert.New(t)

	sumChunker := func(items []sequenceItem) (sequenceItem, Value) {
		sum := 0
		ints := make([]int, len(items))
		for i, item := range items {
			v := item.(int)
			ints[i] = v
			sum += v
		}
		return sum, listFromInts(ints)
	}

	testChunking := func(expect []int, from, to int) {
		seq := newEmptySequenceChunker(sumChunker, sumChunker, modBoundaryChecker{3}, func() boundaryChecker { return modBoundaryChecker{5} })
		for i := from; i <= to; i++ {
			seq.Append(i)
		}

		assert.True(listFromInts(expect).Equals(seq.Done()))
	}

	// [1] is not a chunk boundary, so it won't chunk.
	testChunking([]int{1}, 1, 1)

	// [3] is a chunk boundary, but only a single chunk, so treat it as though it didn't chunk.
	testChunking([]int{3}, 3, 3)

	// None of [1, 2] is a chunk boundary, so it won't chunk.
	testChunking([]int{1, 2}, 1, 2)

	// [3, 4] has a chunk boundary on 3, so it should chunk as [3] [4].
	testChunking([]int{3, 4}, 3, 4)

	// [1, 2, 3] ends in a chunk boundary 3, but only a single chunk, so treat is as though it didn't chunk.
	testChunking([]int{1, 2, 3}, 1, 3)

	// [1, 2, 3, 4] has a chunk boundary on 3, so should chunk as [1, 2, 3] [4].
	testChunking([]int{6, 4}, 1, 4)

	// [1, 2, 3, 4, 5, 6] has a chunk boundary on 3 and 6, so should chunk as [1, 2, 3] [4, 5, 6]. Note also that the level above that ends in a chunk boundary 15 because 4+5+16=15 and 15%5 = 0, but like earlier, this shouldn't chunk because it's only a single chunk.
	testChunking([]int{6, 15}, 1, 6)

	// [1, 2, 3, 4, 5, 6, 7] will chunk as [1, 2, 3] [4, 5, 6] [7] producing meta chunks of [6, 15, 7] which chunks on 15, so we'll get 2 chunks [6, 15] [7].
	testChunking([]int{21, 7}, 1, 7)
}

File: js/src/sequence_chunker_test.js
// @flow

import {suite, test} from 'mocha';
import {assert} from 'chai';
import {int64} from './primitives.js';
import {SequenceChunker} from './sequence_chunker.js';

class ModBoundaryChecker {
  mod: number;

  constructor(mod: number) {
    this.mod = mod;
  }

  get windowSize(): number {
    return 1;
  }

  write(item: int64): boolean {
    return item % this.mod === 0;
  }
}

function sumChunker(items: Array<int64>): [int64, any] {
  let sum = 0;
  for (let i = 0; i < items.length; i++) {
    sum += items[i];
  }

  return [sum, items];
}

suite('SequenceChunker', () => {

  async function testChunking(expect: Array<int64>, from: number, to: number) {
    const seq = new SequenceChunker(null, sumChunker, sumChunker,
      new ModBoundaryChecker(3), () => new ModBoundaryChecker(5));

    for (let i = from; i <= to; i++) {
      seq.append(i);
    }

    assert.deepEqual(expect, await seq.done());
  }

  test('mod', async () => {
    await testChunking([1], 1, 1);
    await testChunking([3], 3, 3);
    await testChunking([1, 2], 1, 2);
    await testChunking([3, 4], 3, 4); // XX
    await testChunking([1, 2, 3], 1, 3);
    await testChunking([6, 4], 1, 4);
    await testChunking([6, 15], 1, 6);
    await testChunking([21, 7], 1, 7);
  });
});
